rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -312,29 +385,121 @@ void GraphCache::parse_put(const std::string & keyexpr)
 ///=============================================================================
 void GraphCache::parse_del(const std::string & keyexpr)
 {
-  // TODO(Yadunund): Validate data.
-  std::vector<std::string> parts = split_keyexpr(keyexpr);
-  if (parts.size() < 3) {
-    RCUTILS_LOG_ERROR_NAMED(
-      "rmw_zenoh_cpp",
-      "Received invalid liveliness token");
+  auto valid_token = _parse_token(keyexpr);
+  if (!valid_token.has_value()) {
+    // Error message has already been logged.
     return;
   }
-  // Get the entity, ie N, MP, MS, SS, SC.
-  const std::string & entity = parts[2];
+  const std::string & entity = valid_token->first;
+  auto node = std::make_shared<GraphNode>(std::move(valid_token->second));
   std::lock_guard<std::mutex> lock(graph_mutex_);
   if (entity == "NN") {
     // Node
+    auto ns_it = graph_.find(node->ns);
+    if (ns_it != graph_.end()) {
+      ns_it->second.erase(node->name);
+    }
     RCUTILS_LOG_WARN_NAMED(
       "rmw_zenoh_cpp",
-      "Removing node %s from the graph.",
-      parts.back().c_str()
+      "Removed node /%s from the graph.",
+      node->name.c_str()
     );
-    graph_.remove(entity.back());
   } else if (entity == "MP") {
     // Publisher
+    if (node->pubs.empty()) {
+      // This should never happen but we make sure _parse_token() has no error.
+      return;
+    }
+    auto ns_it = graph_.find(node->ns);
+    if (ns_it != graph_.end()) {
+      auto node_it = ns_it->second.find(node->name);
+      if (node_it != ns_it->second.end()) {
+        const auto found_node = node_it->second;
+        // Here we iterate throught the list of publishers and remove the one
+        // with matching name, type and qos.
+        // TODO(Yadunund): This can be more optimal than O(n) with some caching.
+        auto erase_it = found_node->pubs.begin();
+        for (; erase_it != found_node->pubs.end(); ++erase_it) {
+          const auto & pub = *erase_it;
+          if (pub.topic == node->pubs.at(0).topic &&
+            pub.type == node->pubs.at(0).type &&
+            pub.qos == node->pubs.at(0).qos)
+          {
+            break;
+          }
+        }
```

Comment:
nit: use `find_if` here and there 
```suggestion
        auto erase_it = std::find_if(found_node->pubs.begin(), found_node->pubs.end(), [&node](const auto &pub) {
            return pub.topic == node->pubs.at(0).topic &&
                   pub.type == node->pubs.at(0).type &&
                   pub.qos == node->pubs.at(0).qos;
        });
```
Then proceed as you do:

```cpp
        if (erase_it != found_node->pubs.end()) {
            // Found a matching element, erase it
            found_node->pubs.erase(erase_it);
            //...
            //...
            //...
        }
```